dmcontent/utils.py

```diff
@@ -9,6 +9,28 @@
 from .errors import ContentTemplateError
 
 
+# jinja's environments are threadsafe (unless you explicitly mutate them during operation, which is not recommended),
+# so it should be safe to keep this as a shared global
+template_environment = DMSandboxedEnvironment(autoescape=True, undefined=StrictUndefined)
+
+
+class _ImmutableTemplateProxy:
+    """
+        A heavily abbreviated proxy for a jinja Template that should ensure immutability and threadsafety
+    """
+    __slots__ = ("render",)
+
+    def __init__(self, template):
+        # instead of keeping a reference to the template on the instance where external code could be e.g. inadvertantly
+        # calling mutating methods on it, seal it in the closure of a function which itself will _become_ our render
+        # "method".
+        self.render = lambda *args, **kwargs: template.render(*args, **kwargs)
+
+    def __deepcopy__(self, memo):
+        # we're immutable.
```

Comment:
Are we actually immutable? What's to stop someone setting `render`?

Reply:
Might be worth changing the comment just so people don't get the wrong idea.

Reply:
Well, there's "immutable" and _immutable_. It's extremely tricky to make a pure-python object actually _immutable_ - `tuple` et al manage it because they're implemented in c, and even then they're not deep-immutable (a `dict` in a `tuple` is still mutable). I _could_ add some more chicanery to make a few more things impossible, but then I would open myself to the accusation that I'm introducing something that's overly-complex and confusing.

It's more about what anyone's ever likely to do without it being obvious that they're being naughty - e.g. it's _possible_ to modify `app.config` from within a flask request, but it's fairly obvious that that would be a bad idea and people using `app.config` don't tend to worry about the possibility that it could change during a request.